griptape/drivers/image_query/openai_vision_image_query_driver.py

```diff
@@ -0,0 +1,56 @@
+from __future__ import annotations
+
+from typing import Literal, Optional
+
+from attr import define, field, Factory
+from openai.types.chat import (
+    ChatCompletionUserMessageParam,
+    ChatCompletionContentPartParam,
+    ChatCompletionContentPartTextParam,
+    ChatCompletionContentPartImageParam,
+)
+import openai
+
+from griptape.artifacts import ImageArtifact, TextArtifact
+from griptape.drivers.image_query.base_image_query_driver import BaseImageQueryDriver
+
+
+@define
+class OpenAiVisionImageQueryDriver(BaseImageQueryDriver):
+    model: str = field(default="gpt-4-vision-preview", kw_only=True, metadata={"serializable": True})
+    api_type: str = field(default=openai.api_type, kw_only=True, metadata={"serializable": True})
+    api_version: Optional[str] = field(default=openai.api_version, kw_only=True, metadata={"serializable": True})
+    base_url: str = field(default=None, kw_only=True, metadata={"serializable": True})
+    api_key: Optional[str] = field(default=None, kw_only=True)
+    organization: Optional[str] = field(default=openai.organization, kw_only=True, metadata={"serializable": True})
+    image_quality: Literal["auto", "low", "high"] = field(default="auto", kw_only=True, metadata={"serializable": True})
+    max_tokens: Optional[int] = field(default=None, kw_only=True, metadata={"serializable": True})
+    client: openai.OpenAI = field(
+        default=Factory(
+            lambda self: openai.OpenAI(api_key=self.api_key, base_url=self.base_url, organization=self.organization),
+            takes_self=True,
+        )
+    )
+
+    def try_query(self, query: str, images: list[ImageArtifact]) -> TextArtifact:
+        message_parts: list[ChatCompletionContentPartParam] = [
+            ChatCompletionContentPartTextParam(type="text", text=query)
+        ]
+
+        for image in images:
```

Comment:
what is the expected behavior for a list of images? does it string all of the text together? wondering why defaulting to multiple?

Reply:
This enables fundamentally different behavior than operating serially over a list of images. As implemented, we can provide many images to Vision and query it in reference to _all the images at once_, not just one at a time. An example would be providing two images and asking the model "What's the difference between these two images?" to understand that one contains a dog and the other contains a cat.

Of course, we can always provide a list of length one to query a single image.